File: packages/newton-dsl-ts/src/task.ts
/**
 * Task handle returned from wf.task(...).
 *
 * Supports chaining: task.then(...).then(...).retry(...)
 */

import { EdgeSpec, PRIORITY_STEP } from "./edges.js";
import { Guard, Ref, makeOutProxy, TaskOutputRef } from "./refs.js";
import type { OutProxy } from "./refs.js";
import type { OperatorCall } from "./operators.js";

export interface RetryOpts {
  times: number;
  waitSeconds?: number;
  multiplier?: number;
  jitterSeconds?: number;
}

export interface ThenOpts {
  when?: Guard | Ref | string;
  label?: string;
}

export class Task {
  public readonly taskId: string;
  public readonly operatorCall: OperatorCall;
  public name: string | null = null;
  public _edges: EdgeSpec[] = [];
  public _nextPriority: number = 0;
  public _retry: Record<string, unknown> | null = null;
  public _maxIterations: number | null = null;
  public _timeoutMs: number | null = null;
  public _terminal: string | null = null;
  public _message: string | null = null;

  constructor(taskId: string, operatorCall: OperatorCall, opts?: { name?: string }) {
    this.taskId = taskId;
    this.operatorCall = operatorCall;
    if (opts?.name) this.name = opts.name;
  }

  // -------------------------------------------------------------------------
  // Transition wiring
  // -------------------------------------------------------------------------

  then(target: Task, opts?: ThenOpts): this {
    let guard: Guard | null = null;
    const w = opts?.when;
    if (w != null) {
      if (w instanceof Guard) {
        guard = w;
      } else if (w instanceof Ref) {
        guard = new Guard(w.rhaiExpr());
      } else {
        guard = new Guard(String(w));
      }
    }

    const edge = new EdgeSpec(
      target.taskId,
      this._nextPriority,
      guard,
      opts?.label ?? null
    );
    this._edges.push(edge);
    this._nextPriority += PRIORITY_STEP;
    return this;
  }

  otherwise(target: Task, opts?: { label?: string }): this {
    const edge = new EdgeSpec(
      target.taskId,
      this._nextPriority,
      null,
      opts?.label ?? null
    );
    this._edges.push(edge);
    this._nextPriority += PRIORITY_STEP;
    return this;
  }

  // -------------------------------------------------------------------------
  // Task configuration
  // -------------------------------------------------------------------------

  retry(opts: RetryOpts): this {
    const r: Record<string, unknown> = {
      max_attempts: opts.times,
      backoff_ms: (opts.waitSeconds ?? 0) * 1000,
    };
    if (opts.multiplier != null) r.backoff_multiplier = opts.multiplier;
    if (opts.jitterSeconds != null) r.jitter_ms = opts.jitterSeconds * 1000;
    this._retry = r;
    return this;
  }

  repeatAtMost(n: number): this {
    this._maxIterations = n;
    return this;
  }

  timeout(seconds: number): this {
    this._timeoutMs = seconds * 1000;
    return this;
  }

  // -------------------------------------------------------------------------
  // Output references
  // -------------------------------------------------------------------------

  get out(): OutProxy {
    return makeOutProxy(this.taskId);
  }

  get output(): TaskOutputRef {
    return new TaskOutputRef(this.taskId);
  }

  // -------------------------------------------------------------------------
  // Serialization
  // -------------------------------------------------------------------------

  toDict(): Record<string, unknown> {
    const d: Record<string, unknown> = {
      id: this.taskId,
      operator: this.operatorCall.operatorType,
    };
    if (this.name != null) d.name = this.name;
    if (this._maxIterations != null) d.max_iterations = this._maxIterations;
    if (this._timeoutMs != null) d.timeout_ms = this._timeoutMs;
    if (this._retry != null) d.retry = this._retry;
    if (this._terminal != null) d.terminal = this._terminal;

    d.params = this.operatorCall.renderedParams();
    d.transitions = this._edges.map((e) => e.toDict());
    return d;
  }
}

File: packages/newton-dsl-py/src/newton/task.py
"""
Task handle returned from wf.task(...).

Supports chaining: task.then(...).then(...).retry(...)
"""
from __future__ import annotations

from typing import TYPE_CHECKING, Any

from .edges import EdgeSpec, PRIORITY_STEP
from .refs import Guard, Ref, _OutAccessor, TaskOutputRef
from .operators import OperatorCall

if TYPE_CHECKING:
    pass


class Task:
    """
    Handle for a workflow task, returned by Workflow.task().
    Stores transitions and configuration that compile.py will serialize.
    """

    def __init__(
        self,
        task_id: str,
        operator_call: OperatorCall,
        *,
        name: str | None = None,
    ) -> None:
        self.task_id = task_id
        self.operator_call = operator_call
        self.name = name
        self._edges: list[EdgeSpec] = []
        self._next_priority: int = 0
        self._retry: dict[str, Any] | None = None
        self._max_iterations: int | None = None
        self._timeout_ms: int | None = None
        self._terminal: str | None = None  # "success" | "failure"
        self._is_terminal: bool = False
        self._message: str | None = None  # for finish/fail tasks

    # ------------------------------------------------------------------
    # Transition wiring
    # ------------------------------------------------------------------

    def then(
        self,
        target: "Task",
        *,
        when: Guard | Ref | str | None = None,
        label: str | None = None,
    ) -> "Task":
        """
        Append a conditional or unconditional transition to `target`.
        Priority is determined by call order (first call = highest priority).
        Returns `self` for fluent chaining.
        """
        from .refs import Guard as _Guard, Ref as _Ref

        guard: _Guard | None = None
        if when is not None:
            if isinstance(when, _Guard):
                guard = when
            elif isinstance(when, _Ref):
                guard = _Guard(when.rhai_expr())
            else:
                guard = _Guard(str(when))

        edge = EdgeSpec(
            target.task_id,
            priority=self._next_priority,
            guard=guard,
            label=label,
        )
        self._edges.append(edge)
        self._next_priority += PRIORITY_STEP
        return self

    def otherwise(self, target: "Task", *, label: str | None = None) -> "Task":
        """
        Append an unconditional fallback transition (lowest priority).
        Returns `self` for fluent chaining.
        """
        # otherwise uses highest declared priority so far + PRIORITY_STEP, no guard
        edge = EdgeSpec(
            target.task_id,
            priority=self._next_priority,
            guard=None,
            label=label,
        )
        self._edges.append(edge)
        self._next_priority += PRIORITY_STEP
        return self

    # ------------------------------------------------------------------
    # Task configuration
    # ------------------------------------------------------------------

    def retry(
        self,
        times: int,
        *,
        wait_seconds: int = 0,
        multiplier: float | None = None,
        jitter_seconds: int | None = None,
    ) -> "Task":
        """Configure retry policy for this task."""
        r: dict[str, Any] = {
            "max_attempts": times,
            "backoff_ms": wait_seconds * 1000,
        }
        if multiplier is not None:
            r["backoff_multiplier"] = multiplier
        if jitter_seconds is not None:
            r["jitter_ms"] = jitter_seconds * 1000
        self._retry = r
        return self

    def repeat_at_most(self, n: int) -> "Task":
        """Set the max_iterations cap (per-task re-entry cap)."""
        self._max_iterations = n
        return self

    def timeout(self, seconds: int) -> "Task":
        """Set task timeout in seconds (stored as timeout_ms)."""
        self._timeout_ms = seconds * 1000
        return self

    # ------------------------------------------------------------------
    # Output reference
    # ------------------------------------------------------------------

    @property
    def out(self) -> _OutAccessor:
        """
        Returns a proxy that turns attribute access into OutRef objects.
        t.out.stdout -> OutRef("tasks.<id>.output.stdout")
        t.out itself renders as TaskOutputRef("tasks.<id>.output") (the whole output)
        """
        known = self.operator_call.output_fields()
        return _OutAccessor(self.task_id, known_fields=known)

    @property
    def output(self) -> TaskOutputRef:
        """
        Returns a reference to the entire task output object.
        t.output -> {"$expr": "tasks.<id>.output"}
        Use when passing the whole output to another operator (e.g., board= param).
        """
        return TaskOutputRef(self.task_id)

    # ------------------------------------------------------------------
    # Serialization helpers
    # ------------------------------------------------------------------

    def to_dict(self) -> dict[str, Any]:
        """Build the task dict for inclusion in the workflow YAML."""
        d: dict[str, Any] = {
            "id": self.task_id,
            "operator": self.operator_call.operator_type,
        }
        if self.name is not None:
            d["name"] = self.name
        if self._max_iterations is not None:
            d["max_iterations"] = self._max_iterations
        if self._timeout_ms is not None:
            d["timeout_ms"] = self._timeout_ms
        if self._retry is not None:
            d["retry"] = self._retry
        if self._terminal is not None:
            d["terminal"] = self._terminal

        params = self.operator_call.rendered_params()
        d["params"] = params

        d["transitions"] = [edge.to_dict() for edge in self._edges]
        return d

    def __repr__(self) -> str:
        return f"Task({self.task_id!r})"
